src/util/io.c

```diff
@@ -146,31 +152,48 @@ socket_connect (
     uint16_t port,
     SOCKET *sock)
 {
-    struct sockaddr_in sockaddr = { 0 };
+    static const struct addrinfo hints = { .ai_socktype = SOCK_STREAM,
+        .ai_family = AF_UNSPEC, .ai_protocol = IPPROTO_TCP,
+        .ai_flags=AI_CANONIDN };
+    struct addrinfo *retp = NULL;
+    struct addrinfo *pprev = NULL;
+    struct addrinfo *p;
+    char port_str[MAX_PORT_STR_LEN];
     int ret = 0;
 
     if (hostname == NULL || sock == NULL) {
         return TSS2_TCTI_RC_BAD_REFERENCE;
     }
 
-    LOG_DEBUG ("Creating AF_INET stream socket");
-    *sock = socket (AF_INET, SOCK_STREAM, IPPROTO_TCP);
-    if (*sock == -1) {
-        LOG_WARNING ("Failed to create socket. errno %d: %s",
-                     errno, strerror (errno));
+    ret = snprintf(port_str, sizeof(port_str), "%u", port);
+    if (ret < 0)
+        return TSS2_TCTI_RC_BAD_VALUE;
+    
+
+    LOG_DEBUG ("Resolving host %s", hostname);
+    ret = getaddrinfo (hostname, port_str, &hints, &retp);
+    if (ret != 0) {
+        LOG_WARNING ("Host %s does not resolve to a valid address: %d: %s",
+            hostname, ret, gai_strerror(ret));
         return TSS2_TCTI_RC_IO_ERROR;
     }
-    sockaddr.sin_family = AF_INET;
-    sockaddr.sin_addr.s_addr = inet_addr (hostname);
-    sockaddr.sin_port = htons (port);
-
-    LOG_DEBUG ("Connecting socket %d to hostname %s on port %" PRIu16,
-               *sock, hostname, port);
-    ret = connect (*sock, (struct sockaddr*)&sockaddr, sizeof (sockaddr));
-    if (ret == -1) {
+    
+    LOG_DEBUG ("Connecting TCP stream socket to host %s on port %" PRIu16,
+               hostname, port);
+    for (p = retp; p != NULL; pprev = p, p = p->ai_next) {
+        *sock = socket (p->ai_family, SOCK_STREAM, 0);
+        if (*sock == -1)
+            continue;
+        if (connect (*sock, p->ai_addr, p->ai_addrlen) != -1)
+            break; // socket connected OK
+        socket_close(sock);
+    }
+    freeaddrinfo (retp);
+    if (p == NULL) {
         LOG_WARNING ("Failed to connect to host %s on port %" PRIu16 ", errno "
-                     "%d: %s", hostname, port, errno, strerror (errno));
-        socket_close (sock);
+                     "%d: %s", (pprev != NULL && pprev->ai_canonname != NULL) ?
+                      pprev->ai_canonname : hostname,
+                      port, errno, strerror (errno));
```

Comment:
For this purpose, I'd say that the flag `AI_CANONNAME` and printing `retp->ai_canonname` are more meaningful.

Of course only before the freeaddrinfo, as it should have been here anyways....